### PR TITLE
Integrate `ferveo` through bindings exposed in `nucypher-core`

### PR DESCRIPTION
**Pull request by piotr-roslaniec**
**Type of PR:**
- Feature

**Required reviews:** 
- 2

**What this does:**
- Integrates pre-release version of `nucypher-core` with `nucypher`
- Replaces `ferveo_py` with Ferveo module exposed from `nucypher.ferveo` 
- Based on changes in https://github.com/nucypher/nucypher-core/pull/53 

**Issues fixed/closed:**
- ...

**Why it's needed:**
- To remove dependency on `ferveo_py` module
- To iron out any issues with `nucypher-core` and `ferveo` before releasing their new versions

**Notes for reviewers:**
- **Please don't merge**
- Are there any other changes that should go into this PR?
- This is a multi-repository effort. Please consider reviewing PRs in other repos:
  - https://github.com/nucypher/ferveo/pull/125
  - https://github.com/nucypher/ferveo/pull/119
  - https://github.com/nucypher/nucypher-core/pull/58

### CODE REVIEW

File: examples/tdec/simple_tdec.py [hunk 1]
```diff
@@ -1,13 +1,12 @@
 import os
-from ferveo_py.ferveo_py import DkgPublicKey
 from pathlib import Path
 
+from nucypher_core.ferveo import DkgPublicKey
```
Comment (KPrasch) [inline examples/tdec/simple_tdec.py:4]: 😎

File: nucypher/crypto/powers.py [hunk 2]
```diff
@@ -205,20 +206,11 @@ class KeyPairBasedPower(CryptoPowerUp):
...
-            # They didn't pass a keypair; we'll make one with the bytes or
-            # Umbral PublicKey if they provided such a thing.
             if public_key:
-                try:
-                    public_key = public_key.as_umbral_pubkey()
-                except AttributeError:
-                    try:
-                        public_key = PublicKey.from_compressed_bytes(public_key)
-                    except TypeError:
-                        public_key = public_key
```
Comment (KPrasch) [inline nucypher/crypto/powers.py:221]: Can you help me understand the rationale for this change?

File: requirements.txt [hunk 3]
```diff
@@ -56,7 +56,7 @@ msgpack==1.0.5
 msgpack-python==0.5.6
 multidict==5.2.0 ; python_version >= '3.6'
 mypy-extensions==0.4.4 ; python_version >= '2.7'
-nucypher-core==0.8.0
+nucypher-core @ git+https://github.com/nucypher/nucypher-core.git@a037b066a7469bccea3b91aee3f092f8de161929#subdirectory=nucypher-core-python
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply requirements.txt:59]: To continue to use a vcs commit as the dependency (from`nucypher-core:main`), please also update the Pipfile entry for `nucypher-core` to be the github dependency before merging this PR.

i.e. use
```
nucypher-core = {git="https://github.com/nucypher/nucypher-core.git", ref="main", subdirectory="nucypher-core-python"}
```

Note using `main` as the ref and not a specific commit in the Pipfile (the specific commit can be used in the requirements file). Doing this allows you to simply relock dependencies, and will update for subsequent relocking of dependencies until we eventually have a release of `nucypher-core` we can use. It's a perpetual reminder to eventually change this to be an actual released version of `nucypher-core`.

File: examples/tdec/simple_tdec.py [hunk 4]
```diff
@@ -1,13 +1,12 @@
...
+
 import nucypher
 from nucypher.blockchain.eth.agents import CoordinatorAgent
 from nucypher.blockchain.eth.registry import LocalContractRegistry
-from nucypher.characters.lawful import Bob
-from nucypher.characters.lawful import Enrico as Enrico
```
Comment (jMyles) [inline examples/tdec/simple_tdec.py:9]: https://getyarn.io/yarn-clip/b56ffaa2-170d-4f03-a606-b8dfb03ddd0c

File: nucypher/crypto/powers.py [hunk 5]
```diff
@@ -253,7 +245,7 @@ class DecryptingPower(KeyPairBasedPower):
 
 class RitualisticPower(KeyPairBasedPower):
     _keypair_class = RitualisticKeypair
-    _default_private_key_class = ferveo_py.Keypair
+    _default_private_key_class = Keypair
```
Comment (KPrasch) [inline nucypher/crypto/powers.py:248]: It's unclear which Keypair this is, even though I know it's intended to be from ferveo.  Let's use a namespace here.

File: nucypher/crypto/powers.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/crypto/powers.py:221]: This intermediate base class is now meant to support two different cryptosystems (umbral and ferveo)  we cannot assume or infer the type.  Might need a little refactoring here.... perhaps the replacing of default keypair classes with `NotImplmplemented`.

Comment (piotr-roslaniec) [reply nucypher/crypto/powers.py:221]: It seems to me that changes to `KeyPairBasedPower` are outside of the scope of this PR. I don't think I have the depth to address them now. I could just revert these changes and document this change in an issue.

Comment (KPrasch) [reply nucypher/crypto/powers.py:221]: 💯

**Review by KPrasch — APPROVED** [review]

**Review by derekpierre — APPROVED** [review]
🎸

**Comment by derekpierre** [comment]
The CI acceptance tests for this PR will fail due to https://github.com/nucypher/nucypher-contracts/pull/84 being merged into `nucypher-contracts:main`. PR #3134 (to be merged afterwards) will fix this issue.

**Comment by piotr-roslaniec** [comment]
@derekpierre Thanks for letting me know, I was convinced some error slipped in

**Comment by derekpierre** [comment]
> @derekpierre Thanks for letting me know, I was convinced some error slipped in

👍 - yeah these are the tricky dependencies.

If you think this PR is ready to be merged, go for it. I can then rebase 3134.

**Review by theref — APPROVED** [review]

**Pull request merged**